### ISSUE TITLE
Common spurious test failures in crater

### ISSUE THREAD

**Issue by saethlin** [sharkdp/fd#919]
Hi! I've been doing some log analysis to try to increase the quality of crater runs. It looks like this crate's integration tests are causing a lot of spurious failures. Close to 50% of the time, the test suite fails. Here are the logs from some recent failing runs: 
https://crater-reports.s3.amazonaws.com/pr-90799/try%238262275854dc0e5381c9d42d56257d94932c5a87/reg/fd-find-8.3.0/log.txt
https://crater-reports.s3.amazonaws.com/beta-1.58-1/1.57.0/reg/fd-find-8.3.0/log.txt
https://crater-reports.s3.amazonaws.com/beta-1.58-1/beta-2021-11-30/reg/fd-find-8.3.0/log.txt

In all these cases, it looks `test_opposing` is failing because the output is in the wrong order. For the life of me I do not know why these tests only fail in crater. I've run them thousands of times, in case this a race condition. I've run the tests in and outside of Docker, and under various filesystem hashing algorithms and I cannot get them to fail locally.

And yet they _are_ failing, and I can't find any handling of unexpected directory iteration order in the failing tests makes me think that's what is going on. Thoughts?

**Comment by tavianator** [comment]
If it's sensitive to directory iteration order then maybe it only fails on certain filesystems.  Do you know what FS crater is using?

**Comment by saethlin** [comment]
I asked in the Zulip and it sounds like it's _probably_ ext4, which is rather frustrating because that's what I can't get them to fail on.

**Comment by tmccombs** [comment]
Is it running in a container? If so, it probably uses overlayfs (with ext4 underneath). Although I also wouldn't expect that to impact the order.

**Comment by saethlin** [comment]
Yes, crater runs in Docker containers. I tried running the tests for this repo in a similar container to what crater uses, couldn't coax a test failure out of it.

**Comment by tmccombs** [comment]
Ok, I think I have it. That test runs two commands and test if it has the same output. Normally, the result will be sorted, because the results will be found fast enough that we finish before filling the buffer or reaching the timeout. However, on a crater run it may not have as many resources available, so it takes longer, and ends up timing out and not sorting.

I think #937 should fix it

**Comment by saethlin** [comment]
> However, on a crater run it may not have as many resources

Ah! This sounds exactly right. Most crater runs end up incredibly memory-bound, and I don't know how cgroup memory limits work, but time in the kernel goes through the roof for many minutes when memory gets tight.

**Comment by tmccombs** [comment]
That PR has been merged, let us know if it continues to be flaky or not.

**Comment by tavianator** [comment]
I couldn't find a list of recent *completed* crater runs anywhere, but I did find this one by looking for @craterbot's comments: https://crater-reports.s3.amazonaws.com/beta-1.61-1/1.60.0/gh/sharkdp.fd/log.txt

I think it's likely this is now fixed, but feel free to re-open or file a new issue if there are more issues with crater.
